Action to run accessibility tests and create markdown report

gabalafou opened the pull request:
The major parts in this PR are:

- An action to build JupyterLab from source and run accessibility regression tests against it using Playwright
- A workflow to dispatch that action on demand
- A custom Playwright reporter (markdown) that outputs the Playwright test results to a markdown file
- Separating the one regression test (no tab traps in initial page) from the axe-core tests into a new directory called `regression-tests`
- Creating a new directory called `manual-testing-scripts` to hold reference markdown files that describe each accessibility test. Inspiration: [ACT Rules](https://act-rules.github.io/rules/). 
- Adding a manual testing script for the no-tab-traps-initial-page test 
- Providing a way to link to the manual testing scripts from the markdown report of test results

gabalafou committed:
Build JupyterLab workflow

gabalafou committed:
runs-on is required

gabalafou committed:
fix job id

gabalafou committed:
test jupyterlab from source

gabalafou committed:
copy paste

gabalafou committed:
no idea what im doing

gabalafou committed:
do not use base setup

gabalafou committed:
do not use conda prefix

Commit:
some more tries

Commit:
add workflow_call event

Commit:
add workflow as action

Commit:
update action.yml

Commit:
update action and workflow

Commit:
use github and json reporters

Commit:
correct path to environment file

Commit:
correct path to artifacts

Commit:
upload playwright report

Commit:
jupyter lab build and --with-deps

Commit:
dev build but why?

Commit:
put jlpm run build in the right place

Commit:
markdown reporter and github summary

Commit:
.md not .ts

Commit:
use older version of outputReport function

Commit:
strip ansi escapes

Commit:
update markdown reporter

Commit:
update action

Commit:
add correct working-directory in action

Commit:
update workflow

Commit:
remove build-jupterlab, rewrap markdown

Commit:
fixes

Commit:
template

Commit:
checkout in the workflow not the action

Commit:
fix paths

Commit:
correct path to the test result files

gabalafou commented:
OMG it worked! You can see it in the checks above, or by looking at [the last run for the jlab-a11y workflow](https://github.com/Quansight-Labs/jupyter-a11y-testing/actions/runs/3296312648).

gabalafou commented:
Almost. The test info link in the GitHub job summary is broken - see [the raw summary](https://github.com/Quansight-Labs/jupyter-a11y-testing/actions/runs/3296312648/jobs/5435826540/summary_raw). Need to figure out how to get that right.

Commit:
fix markdown link

Commit:
fix paths again

gabalafou commented on testing/jupyterlab/markdown-reporter.ts:
```diff
@@ -0,0 +1,130 @@
+// my-awesome-reporter.ts
```
```suggestion
// markdown-reporter.ts
// 
// Output Playwright test results in Markdown format.
// This code is basically a copy-paste job with modifications
// from Playwright's built-in JSON reporter, available on GitHub at
// https://github.com/microsoft/playwright/blob/main/packages/playwright-test/src/reporters/json.ts
```

gabalafou commented on .github/actions/jlab-a11y-test/action.yml:
```diff
@@ -0,0 +1,116 @@
+      shell: bash
+
+    - name: Cache conda 🧠
+      uses: actions/cache@v2
```
```suggestion
      uses: actions/cache@v3
```

We'll want to update this if we want to stop getting warnings from GitHub about Node 12 being deprecated soon.

gabalafou commented on .github/actions/jlab-a11y-test/action.yml:
```diff
@@ -0,0 +1,116 @@
+
+    - name: Build JupyterLab
+      run: |
+        conda run --prefix $CONDA_PREFIX ./scripts/ci_install.sh
```
It seems to me that ideally we would create two separate Conda environments, one for JupyterLab and one for our testing application. But I'm not sure how to do that.

trallard replied:
I am not sure I follow the reasoning for this - having two separate environments with packages and dependencies that are needed to complete the testing sounds like an unnecessarily complex setup (debugging, dependencies mismatch, and calling the correct processes)

steff456 replied:
Is there any way you could test the app without JupyterLab? That way you can have a test with JyputerLab and another one with just the app.

trallard replied:
there is not an application _per se_ - what Gabriel refers to as an "app" here is the collection of Playwright tests

gabalafou commented on .github/actions/jlab-a11y-test/action.yml:
```diff
@@ -0,0 +1,116 @@
+        echo $CONDA_PREFIX
+      shell: bash -l {0}
+
+    - name: Checkout JupyterLab
```
I wonder if I should pull the steps that checkout, build and install JupyterLab into a separate action that this action calls.

trallard replied:
there is some refactor needed - let's add any improvements for later PR's

gabalafou commented on testing/jupyterlab/playwright.config.ts:
```diff
@@ -13,10 +13,10 @@ expect.extend(matchers);
   ...galataConfig,
 
-  testDir: "./tests",
+  testDir: "./regression-tests",
```
@trallard remember how we talked about how the testing work probably has two separate branches?

- regression test suite
- accessibility checks for things like weekly testing, automated reports, etc.

I decided to separate the one regression test (no tab traps) from the other test (run Axe-core checks), keeping them in separate directories like so:

- Axe-core checks -> `tests/`
- regression tests -> `regression-tests/`

While putting them in separate directories makes the separation very clear, I wonder if this is not the right approach. Specifically I'm wondering if I should use [Playwright test tags](https://playwright.dev/docs/test-annotations#tag-tests) instead. Or perhaps even another approach.

The reason I think that maybe they should all live in the same folder is that while the regression tests are their own thing, they should also be wrapped up into the bigger test report.

trallard replied:
I'd argue all tests should be under a `tests` dir and then into subdirs based on the scope

Annotations sound like a good approach for cherry-picking tests

gabalafou replied:
I'll change it back to all tests living in the tests directory.

gabalafou commented on testing/jupyterlab/package.json:
```diff
@@ -11,7 +11,8 @@
   "dependencies": {
-    "@jupyterlab/galata": "4.0.2",
+    "@jupyterlab/galata": "4.3.8",
+    "@playwright/test": "^1.27.1",
```
I added Playwright as a direct dependency because I want to be able to use the latest Playwright features without being stuck with whatever version of Playwright comes along with latest release of Galata.

gabalafou commented on testing/jupyterlab/markdown-reporter.ts:
```diff
@@ -0,0 +1,130 @@
+      if (errorLines.length > 0) {
+        lines.push(`${indent}<details><summary>`);
+      }
+      lines.push(`${indent}Error: ${stripAnsiEscapes(firstLine)}`);
```
Need to add comments to this file, such as why I'm calling this `stripAnsiEscapes` function.

gabalafou commented on testing/jupyterlab/jupyter_server_config.py:
```diff
@@ -8,6 +8,7 @@
 c.ServerApp.port = 8888
 c.ServerApp.port_retries = 0
 c.ServerApp.open_browser = False
+c.LabApp.dev_mode = True
```
I wish I knew a way to test JupyterLab from source NOT having to use dev mode. Because when I test JupyterLab, I want it to be as close as possible to what end users use.

trallard replied:
if running regression tests _within a development workflow_, then installing in `dev` mode is a natural choice since it will link and use any unpublished JS packages 

_as close to users_ would fall perhaps into _acceptance testing_ or _e2e_ (can think of adding these at pre-release, release, or  on a chron job

gabalafou commented on testing/jupyterlab/test-info-pages/no-tab-trap-initial-page.md:
```diff
@@ -0,0 +1,38 @@
+# No Tab Trap on Initial Page
```
My vision here is to a have a Markdown "info page" for every regression test, explaining what fix it checks, why that fix is important, what accessibility criteria it relates to, how to perform the test manually, etc. 

I want something like the [ACT Rule Template](https://act-rules.github.io/pages/design/rule-template/) but I want the format to be a bit more lightweight, if possible, so that it's not too burdensome to add a new regression test. I'm thinking that the final format will become evident after we've added about 10 new tests to the repo.

gabalafou commented on testing/jupyterlab/test-info-pages/no-tab-trap-initial-page.md:
```diff
@@ -0,0 +1,38 @@
+
+If this test fails, it means the app fails to conform to 2.1.2.
+
+If this test passes, however, it does not mean that the app conforms to 2.1.2.
```
I think most of the regression tests will be like this, i.e., they can tell you if you've violated a WCAG criterion but not if the entire app conforms to the criterion, so we might not need to explain this in every test info page. We might need to have one page of definitions and explanations that each individual test info page links back to.

trallard replied:
I find the whole way this is phrased a tad confusing:

- Fail -> no conformance
- Pass -> still no conformance? what does this mean concerning the criterion? what can I verify then?

gabalafou replied:
Yeah that's the thing about many of the automated tests: they can only tell you if you fail to conform to WCAG.

But I'll reword this section to be less abstract.

gabalafou commented on testing/jupyterlab/test-info-pages/no-tab-trap-initial-page.md:
```diff
@@ -0,0 +1,38 @@
+# No Tab Trap on Initial Page
```
I wasn't sure what to call these documents. I didn't really like the name "testing scripts" because "scripts" usually means a code file that automates a task, which is the opposite of how we mean it here.

Some alternatives I considered:

- test docs, which I didn't like because I think people might think that they are the docs for the repo or for the whole testing apparatus
- test wiki pages - no because I don't think these are really a wiki, the scope is too narrow
- test descriptions - this is a pretty accurate name but I worry that it's not specific enough because there are other things in the system that somebody might refer to as "test descriptions" for example the text in `test.describe` blocks. I want these things to be called something that when people say the word, they know you're talking about this and not that.
- test summaries - but then I think people will think these are about the results of test output
- test resumes - I kind of liked this one but I also worried if the word resume is too close to summary. Like, again, I don't want folks thinking that this is about summarizing the outputs of the test runs.
- test profiles - but then devs might think we're talking about code performance profiling (CPU, memory usage)

isabela-pf replied:
I don't have a strong opinion, but I appreciate that you documented possible confusion points. Info and/or info pages works for me.

trallard replied:
these seem to fall under manual testing scripts (which is the official and industry-wide adopted term)
it is often better to look at what the standard is than creating a new pattern and accounting for all possible confusion from folks

gabalafou replied:
Oh interesting! I didn't know that. @trallard could you point me to a variety of bodies of manual testing scripts? That way I can check if the format of those is close enough to the format of what I'm proposing

trallard replied:
I searched for some good examples with no avail - since most of these tests I have seen are internal documents, I have no real-world examples to share. But generally, they have sections like

- Metadata: author, date updated, version, etc.
- Description: what this is, what does it test, and any compliance/requirements needed
- A table with info like
  - Steps (instructions)
  - Inputs (if any)
  - Expected outcome

gabalafou commented on .github/actions/jlab-a11y-test/action.yml:
```diff
@@ -0,0 +1,116 @@
+  against it using Playwright. Outputs the results to a markdown file which is
+  then copied to the GitHub workflow run summary.
+inputs:
+  repository:
+    description: JupyterLab repository, for example gabalafou/jupyterlab
+  ref:
```
The names of these inputs are copied from actions/checkout because they are passed directly through to that action.

gabalafou commented on testing/jupyterlab/markdown-reporter.ts:
```diff
@@ -0,0 +1,130 @@
+    const [_root, _project, file, ...titlePaths] = test.titlePath();
+    const title = titlePaths.join(" > ");
+    const infoURL = this.getTestInfoURL(test);
+    lines.push(`- ${outcome} [${title}](${infoURL})`);
```
I'm not super happy right now with the way the Markdown report looks visually, but it can be improved later.

I also wonder if perhaps I should use a templating language like Handlebars to make it easier to edit the way the report looks. I mean, right now it's not so easy to get a sense of how the entire Markdown file will look just from looking at this file.

trallard replied:
is the tab trap test the only one running atm? 

again depending on how we add/extend tests, this can be extended; for now, this should be more than enough

that said - I feel some metadata could be included in the template/markdown

gabalafou replied:
Metadata.... such as?

gabalafou commented on .github/workflows/jlab-a11y-test.yml:
```diff
@@ -16,63 +33,29 @@ concurrency:
-      - name: Upload testing results 📤
-        uses: actions/upload-artifact@v3
-        # using default retention policy = 90 days
+      - name: Run testing action
```
If we want to be able to run the testing action from the JupyterLab repo, I need to put the checkout step above into the action itself, but I'm not quite sure how to express the following in a workflow: checkout the repo that this action belongs to, **not** the repo that called this action.

trallard replied:
we can do this in a later iteration

gabalafou reviewed (commented):
Add some code comments to help explain my reasoning and thought processes behind the code.

isabela-pf commented on testing/jupyterlab/test-info-pages/no-tab-trap-initial-page.md:
```diff
@@ -0,0 +1,38 @@
+## Description
+
+When a user loads the app for the first time, they should be able to use the tab
+key to navigate without getting trapped.
```
```suggestion
key to navigate the page they land on without getting trapped.
```

isabela-pf commented on testing/jupyterlab/test-info-pages/no-tab-trap-initial-page.md:
```diff
@@ -0,0 +1,38 @@
+
+If this test passes, however, it does not mean that the app conforms to 2.1.2.
+It only means that the app does not violate 2.1.2 for this specific test. This
+test can be combined with others to increase confidence that the app as a whole
```
```suggestion
test can be combined with tab trap tests for other pages or page states to increase confidence that the app as a whole
```

If the intent is to have this as a general disclaimer, then I think what you have works. If you want to give an idea on this test specifically, I think it would be nice to get a hint of why a passing test does not equate to conformance and/or what would be more thorough.

I'm trying to think of this as if I'm someone without a lot of past web accessibility experience, and I could imagine myself being like "well, what does the test do if it doesn't tell me if I am complying with the standard?" So I imagine pointing out the "why" (like how this only tests one page of a whole app) could help.

gabalafou replied:
Oh yes, I like this suggestion. We can have some general abstract language somewhere else but in the test info page we should make it as concrete as possible.

trallard replied:
aha Isabela added on my thoughts!

What I want to know is: what does this test tell me? what have I learned?

gabalafou replied:
👍 love this, will reword

isabela-pf commented on testing/jupyterlab/test-info-pages/no-tab-trap-initial-page.md:
```diff
@@ -0,0 +1,38 @@
+test can be combined with others to increase confidence that the app as a whole
+conforms to 2.1.2.
+
+## How to Verify the Test Manually
```
I really like this section! Out of curiosity, are you basing this on any manual testing resource, or is this just a reflection of how you learned to test it?

gabalafou replied:
I'm not quite sure I understand what you're asking. Are you asking specifically about the technique I'm using to test for no tab traps (i.e., cycle back to the beginning) or are you asking about the way I wrote it up?

In either case, it's all my own work. I don't recall getting any of this from anywhere else.

isabela-pf reviewed (approved):
I know I left some comments, but there's nothing here that I would be uncomfortable merging as-is in `testing/jupyterlab/test-info-pages/no-tab-trap-initial-page.md`. As such, I'm happy to approve. 

It's good to see these tests up; thank you so much!

trallard commented on .github/actions/jlab-a11y-test/action.yml:
```diff
@@ -0,0 +1,116 @@
+  Builds JupyterLab from source, then runs accessibility regression tests
+  against it using Playwright. Outputs the results to a markdown file which is
+  then copied to the GitHub workflow run summary.
+inputs:
```
```suggestion
# see https://github.com/actions/checkout more details on the inputs accepted
inputs:
```

gabalafou commented on testing/jupyterlab/test-info-pages/no-tab-trap-initial-page.md:
```diff
@@ -0,0 +1,38 @@
+test can be combined with others to increase confidence that the app as a whole
+conforms to 2.1.2.
+
+## How to Verify the Test Manually
```
```suggestion
## How to Perform the Test Manually
```

trallard commented on .github/actions/jlab-a11y-test/action.yml:
```diff
@@ -0,0 +1,116 @@
+  steps:
+    - name: List environment variables
+      run: env
+      shell: bash -l {0}
```
Does this need to be executed as a `login` shell? 
There is no need to execute any startup files to access any of the GH environment variables

```suggestion
```

trallard commented on .github/actions/jlab-a11y-test/action.yml:
```diff
@@ -0,0 +1,116 @@
+      run: env
+      shell: bash -l {0}
+
+    - run: echo "🎉 The job was automatically triggered by a ${{ github.event_name }} event."
+      shell: bash
+    - run: echo "🐧 This job is now running on a ${{ runner.os }} server hosted by GitHub!"
+      shell: bash
```
I do not think this is needed at all - these actions will always be run in a GH action server and having this information displayed seems redundant

gabalafou replied:
Agreed. I'll remove it later. I don't want to remove debugging info yet while this is still in early development.

gabalafou replied:
Would be nice if GitHub had better debug and introspection features so I don't have to do stuff like this.

trallard replied:
have you tried looking at the workflow debugging logs?

https://docs.github.com/en/actions/monitoring-and-troubleshooting-workflows/using-workflow-run-logs#viewing-logs-to-diagnose-failures

trallard commented on .github/actions/jlab-a11y-test/action.yml:
```diff
@@ -0,0 +1,116 @@
+    - run: echo "🎉 The job was automatically triggered by a ${{ github.event_name }} event."
+      shell: bash
+    - run: echo "🐧 This job is now running on a ${{ runner.os }} server hosted by GitHub!"
+      shell: bash
```
Also, there is no need to set the shell at every step - bash is the default shell on any UNIX-based runner. So this needs fixing throughout

```suggestion
```

gabalafou replied:
Hmm, I seem to remember something complaining when I removed `shell` from each step in the action file. If I remember correctly, this is a discrepancy between workflow and action files. The action file needs to specify the shell at each step but the workflow does not. However, I cannot recall at the moment where I saw this, if it was the VSCode GitHub action linter extension, which isn't totally reliable, or if it was GitHub that complained.

gabalafou replied:
Okay. It looks like GitHub definitely requires the `shell` property for each step in an action:

```
Error: GitHub.DistributedTask.ObjectTemplating.TemplateValidationException: The template is not valid. 
/home/runner/work/jupyter-a11y-testing/jupyter-a11y-testing/./.github/actions/jlab-a11y-test/action.yml 
(Line: 15, Col: 7): Required property is missing: shell
```

gabalafou replied:
For each step that uses `run`, that is. (Steps that don't use `run`, for example steps that call other actions, do not need to specify `shell`.)

trallard replied:
ah yes this is true for actions - I forgot as I was doing the review. 

also I'd suggest giving this file a better name like `setup-tests` or something else that gives an indication of what is happening here

trallard commented on .github/actions/jlab-a11y-test/action.yml:
```diff
@@ -0,0 +1,116 @@
+    - run: echo "🔎 The name of your branch is ${{ github.ref }} and your repository is ${{ github.repository }}."
+      shell: bash
+    - run: |
+        echo "github.workspace = ${{ github.workspace }}"
+        echo "github.sha = ${{ github.sha }}"
+        echo "github.repository = ${{ github.repository }}"
+        echo "github.server_url = ${{ github.server_url }}"
+      shell: bash
```
This again seems redundant - there is a lot of unnecessary stdout for redundant information

gabalafou replied:
Agreed. I'll remove it later. I don't want to remove debugging info yet while this is still in early development.

trallard commented on .github/actions/jlab-a11y-test/action.yml:
```diff
@@ -0,0 +1,116 @@
+      run: |
+        which jupyter
+        jupyter lab --version
+      shell: bash -l {0}
```
same comment - do we need to set an interactive bash?

gabalafou replied:
For this step, I think so? Otherwise I'm not sure the shell will be activated to the right conda environment. I got this from your code. There's a comment in the docs for the conda-incubator/setup-miniconda action about this.

I'd have to remove this line though, and run the workflow to be sure. 

Honestly, shell scripting is not my forte 😅.

trallard commented on .github/workflows/jlab-a11y-test.yml:
```diff
@@ -16,63 +33,29 @@ concurrency:
+      - run: echo "github.workspace = ${{ github.workspace }}"
+      - run: echo "github.sha = ${{ github.sha }}"
+      - run: echo "github.repository = ${{ github.repository }}"
+      - run: echo "github.server_url = ${{ github.server_url }}"
```
Same comment as above. This whole job seems redundant - looking at https://github.com/Quansight-Labs/jupyter-a11y-testing/actions/runs/3301320792/jobs/5446659566 

a whole lot of the info is missing and therefore think it does not add loads of value

trallard commented on .github/actions/jlab-a11y-test/action.yml:
```diff
@@ -0,0 +1,116 @@
+        echo "sha8=$(echo ${GITHUB_SHA} | cut -c1-8)" >> $GITHUB_OUTPUT
+      shell: bash -l {0}
+
+    - name: Upload Playwright report 📤
+      uses: actions/upload-artifact@v3
+      # using default retention policy = 90 days
+      with:
+        name: playwright-report-${{ steps.getrefs.outputs.branch }}-${{ steps.getrefs.outputs.sha8 }}
+        path: ./testing/jupyterlab/playwright-report/
+
+    - name: Upload JSON test results 📑
+      uses: actions/upload-artifact@v3
+      with:
+        name: json-results-test-a11y-${{ steps.getrefs.outputs.branch }}-${{ steps.getrefs.outputs.sha8 }}
+        path: ./testing/jupyterlab/test-results/jupyterlab-a11y-regression-test-results.json
```
is it best to have two separate artefacts? or would it make more sense to have only one zip for folks to download if needed

gabalafou replied:
I think it would be good to create a single zip file with a README. I'll do that in a follow-up:

- https://github.com/Quansight-Labs/jupyter-a11y-testing/issues/26

gabalafou replied:
- https://github.com/Quansight-Labs/jupyter-a11y-testing/issues/26

trallard commented on testing/jupyterlab/markdown-reporter.ts:
```diff
@@ -0,0 +1,130 @@
+  Reporter,
+} from "@playwright/test/types/testReporter";
+
+class MyReporter implements Reporter {
```
Suggestion: replace `MyReporter` with a more meaningful name

trallard commented on testing/jupyterlab/playwright.config.ts:
```diff
@@ -39,7 +39,21 @@ const config: PlaywrightTestConfig = {
+        outputFile: "test-results/jupyterlab-a11y-regression-test-results.md",
+        /* The URL to the test info pages directory */
+        testInfoPagesBaseURL: process.env.CI
+          ? `${process.env.GITHUB_SERVER_URL}/${process.env.GITHUB_REPOSITORY}/blob/${process.env.GITHUB_SHA}/testing/jupyterlab/test-info-pages`
+          : "https://github.com/Quansight-Labs/jupyter-a11y-testing/blob/main/testing/jupyterlab/test-info-pages",
```
Is this so that it always points to the file within the PR?

gabalafou replied:
Yes, line 50 should allow the GitHub Action to run against a PR that adds a new test + test info page (you should never add a test without adding a test info page). It should allow it to create a markdown report in the GitHub workflow run summary in which the link to the test info page isn't broken. The link to the test info page in the summary should take you to the info page in the working branch of the PR.

gabalafou replied:
I'll add a comment explaining the purpose of this line.

trallard commented on testing/jupyterlab/regression-tests/no-tab-trap-initial-page.test.ts:
```diff
@@ -14,6 +14,10 @@ import { test } from "@jupyterlab/galata";
   test("should cycle through elements back to the start", async ({ page }) => {
+    test.info().annotations.push({
+      type: "link",
+      description: "/no-tab-trap-initial-page.md"
```
When looking at the HTML report this "link" displays the bare file name is this intended? or was it intended to be a clickable link to the URL?

gabalafou replied:
Yeah I noticed that after the fact. This needs a rethink.

trallard replied:
let's make an issue to fix this later

trallard commented on testing/jupyterlab/test-info-pages/no-tab-trap-initial-page.md:
```diff
@@ -0,0 +1,38 @@
+# No Tab Trap on Initial Page
+
+- id: 1wlgh6
```
where is this id created from?

gabalafou replied:
[random.org/strings](https://www.random.org/strings/?num=1&len=6&digits=on&loweralpha=on&format=html&rnd=new)

ACT [uses random ids](https://act-rules.github.io/pages/design/rule-design/) for their rules. That way the name of the test can change but it keeps the same id.

gabalafou committed:
Apply suggestions from code review

Co-authored-by: Tania Allard <taniar.allard@gmail.com>
Co-authored-by: Isabela Presedo-Floyd <50221806+isabela-pf@users.noreply.github.com>

Commit:
Put regressions under sub-directory of tests

Commit:
reword test info page

Commit:
add shell back

Commit:
rename action and workflow files

Commit:
fix syntax error

Commit:
a few more finishing touches

Commit:
really need a linter

Commit:
fix default export

Commit:
Review driven changes

- Update docs
- Update workflow dispatch to accept test project as input
- Rename test info pages to manual testing scripts
- Add comment to markdown reporter

Commit:
fix name workflow file

Commit:
forgot to rename a few places

trallard reviewed (commented):
Merging now and will deal with improvements / extras in subsequent iterations
